src/axes/axis.ts

```diff
@@ -150,9 +168,216 @@ export class Axis<D> extends Component {
+      .attr("r", 3)
+      .attr("visibility", visibilityF);
+
+    let rectangleOffsetF = (d: D) => {
+      switch (this.orientation()) {
+        case "bottom":
+        case "right":
+          return offsetF(d);
+        case "top":
+        case "left":
+          return offsetF(d) - measurements.get(d).height;
+      }
+    };
+    bindElements(this._annotationContainer.select(".annotation-rect-container"), "rect", "annotation-rect")
+      .attr(this._isHorizontal() ? "x" : "y", positionF)
+      .attr(this._isHorizontal() ? "y" : "x", rectangleOffsetF)
+      .attr(this._isHorizontal() ? "width" : "height", (d) => measurements.get(d).width)
+      .attr(this._isHorizontal() ? "height" : "width", (d) => measurements.get(d).height)
+      .attr("visibility", visibilityF);
+
+    let annotationWriter = this._annotationWriter;
+    let annotationFormatter = this.annotationFormatter();
+    let isHorizontal = this._isHorizontal();
+    bindElements(this._annotationContainer.select(".annotation-label-container"), "g", "annotation-label")
+      .attr("transform", (d) => {
+        let xTranslate = this._isHorizontal() ? positionF(d) : rectangleOffsetF(d);
+        let yTranslate = this._isHorizontal() ? rectangleOffsetF(d) : positionF(d);
+        return "translate(" + xTranslate + "," + yTranslate + ")";
+      })
+      .attr("visibility", visibilityF)
+      .each(function (annotationLabel) {
+        let writeOptions = {
+          selection: d3.select(this),
+          xAlign: "center",
+          yAlign: "center",
+          textRotation: isHorizontal ? 0 : 90
+        };
+        annotationWriter.write(annotationFormatter(annotationLabel),
+                                 isHorizontal ? measurements.get(annotationLabel).width : measurements.get(annotationLabel).height,
+                                 isHorizontal ? measurements.get(annotationLabel).height : measurements.get(annotationLabel).width,
+                                 writeOptions);
+      });
+  }
+
+  protected _axisHeightWithoutMargin() {
+    let relevantDimension = this._isHorizontal() ? this.height() : this.width();
+    let axisHeightWithoutMargin = this._isHorizontal() ? this._computedHeight : this._computedWidth;
+    return Math.min(axisHeightWithoutMargin, relevantDimension);
+  }
+
+  private _annotationToTier(measurements: Utils.Map<D, SVGTypewriter.Measurers.Dimensions>) {
+    let annotationTiers: D[][] = [[]];
+    let annotationToTier = new Utils.Map<D, number>();
+    let dimension = this._isHorizontal() ? this.width() : this.height();
+    this.annotatedTicks().forEach((annotatedTick) => {
+      let position = this._scale.scale(annotatedTick);
+      let length = measurements.get(annotatedTick).width;
+      if (position < 0 || position + length > dimension) {
+        annotationToTier.set(annotatedTick, -1);
+        return;
+      }
+      let tierHasCollision = (testTier: number) => annotationTiers[testTier].some((testTick) => {
```

Comment:
`n` is likely small enough that it will not be relevant (text takes up a lot of space inherently, which means the idea of even > 10 annotations cannot give a really good result)